File: algorithm/avl_tree.cpp
/**
 * 平衡二叉树
 * avl tree
 */
#include <iostream>
#include <stdlib.h>

using namespace std;

struct avl_node
{
    avl_node():data(0), height(0), freq(1), left(NULL), right(NULL) {}
    int data;
    int height;
    int freq;       // 表示相同的此结点有几个了
    avl_node* left;
    avl_node* right;
};

class avl_tree
{
public:
    avl_tree():root(NULL){}
    void insertData(int data)  //  插入
    {
        insertNode(root, data);
    }
    void deleteData(int data)  //  删除
    {
        deleteNode(root, data);
    }
    void traverse()            //  遍历
    {
        cout<<"avl "<<root->height<<": ";
        inOrderTraverse(root);
        cout<<endl;
    }
    avl_node* find(int data)   //  查找
    {
        return findNode(root, data);
    }
private:
    avl_node* root;
    void deleteNode(avl_node* &root, int data)
    {
        if(root == NULL)
            return;
        if(root->data > data)
        {
            deleteNode(root->left, data);
            if(2==height(root->left)-height(root->right))
            {
                if(data < root->left->data)
                    root = SingRotateLeft(root);
                else
                    root = DoubleRotateLR (root);
            }
        }
        else if(root->data < data)
        {
            deleteNode(root->right, data);
            if(2==height(root->left)-height(root->right))
            {
                if(data > root->right->data)
                    root = SingRotateRight(root);
                else
                    root = DoubleRotateRL(root);
            }
        }
        else // 若相等，则删除此结点
        {
            if(root->left && root->right)
            {
                avl_node* tmp = root->right;//tmp指向结点右儿子
                while(tmp->left)//找到右子树中的最小结点
                    tmp = tmp->left;
                root->data = tmp->data;
                root->freq = tmp->freq;
                deleteNode(root->right, tmp->data);
                if(2==height(root->left)-height(root->right))
                {// 当结点删除的时候，root的右子树只会降低，所以，左-右==2，则旋转
                    if(root->left->right!=NULL&& (height(root->left->right)>height(root->left->left) ))
                        // 如果root->left的右子树更高，则LR旋转，否则左旋
                        root = DoubleRotateLR(root);
                    else
                        root = SingRotateLeft(root);
                }
            }
            else
            {
                avl_node* tmp = root;
                if(root->left)
                    root = root->left;
                else if(root->right)
                    root = root->right;
                else
                    root = NULL;
            }
        }

        if(root)
            root->height = Max(height(root->left), height(root->right)) + 1;
    }
    void insertNode(avl_node* &root, int data)
    {
        if(root == NULL)
        {
            root = new avl_node;
            root->data = data;
            return;
        }
        if(root->data > data)
        {
            insertNode(root->left, data);
            if(2==height(root->left)-height(root->right))
            {
                if(data < root->left->data)
                    root = SingRotateLeft(root);
                else
                    root = DoubleRotateLR (root);
            }
        }
        else if(root->data < data)
        {
            insertNode(root->right, data);
            if(2==height(root->left)-height(root->right))
            {
                if(data > root->right->data)
                    root = SingRotateRight(root);
                else
                    root = DoubleRotateRL(root);
            }
        }
        else
            root->freq++;

        root->height = Max(height(root->left), height(root->right)) + 1;
    }

    avl_node* SingRotateLeft(avl_node* &k2)  //左左情况下的旋转
    {// 左左情况下，向右旋转
        avl_node* k1 = k2->left;
        k2->left = k1->right;
        k1->right = k2;

        k2->height = Max(height(k2->left), height(k2->right)) + 1;
        k1->height = Max(height(k1->left), height(k1->right)) + 1;
        return k1;
    }
    avl_node* SingRotateRight(avl_node* &k2) //右右情况下的旋转
    {// 右右情况下，向左旋转
        avl_node* k1 = k2->right;
        k2->right = k1->left;
        k1->left = k2;

        k2->height = Max(height(k2->left), height(k2->right)) + 1;
        k1->height = Max(height(k1->left), height(k1->right)) + 1;
        return k1;
    }
    avl_node* DoubleRotateLR(avl_node* &k3)  //左右情况下的旋转
    {
        k3->left = SingRotateRight(k3->left);
        return SingRotateLeft(k3);
    }
    avl_node* DoubleRotateRL(avl_node* &k3)  //右左情况下的旋转
    {
        k3->right = SingRotateLeft(k3->right);
        return SingRotateRight(k3);
    }

    int Max(int a, int b) {return a>b?a:b;}
    avl_node* findNode(avl_node* root, int data)
    {
        if(!root)
            return NULL;
        if(root->data > data)
            return findNode(root->left, data);
        else if(root->data < data)
            return findNode(root->right, data);
        else
            return root;
    }
    int height(avl_node* root)
    {
        if(!root)
            return 0;
        return root->height;
        //return Max(height(root->left), height(root->right)) + 1;
    }

    void inOrderTraverse(avl_node* root)
    {
        if(!root)
            return;
        inOrderTraverse(root->left);
        cout<<root->data<<" ";
        inOrderTraverse(root->right);
    }

};

int main()
{
    avl_tree tree;
    for(int i=0; i<10; ++i)
    {
        tree.insertData(rand()%10);
        tree.traverse();
    }
    for(int i=0; i<10; ++i)
    {
        tree.deleteData(rand()%10);
        tree.traverse();
    }
    return 0;
}

File: javaLearn/src/main/java/org/albon/java/learn/avl/tree/AvlTree.java
package org.albon.java.learn.avl.tree;

public class AvlTree {

    private TreeNode root = null;

    public void insertData(int data) {
        this.root = insertNode(this.root, data);
    }

    public void traverse() {
        System.out.println("traverse start: ");
        inOrderTraverse(root);
        System.out.println("traverse end: ");
    }

    public void deleteData(int data) {
        this.root = deleteNode(this.root, data);
    }

    private TreeNode deleteNode(TreeNode root, int data) {
        if (root == null) {
            return null;
        }

        if (root.getData() > data) {
            root.setLeft(deleteNode(root.getLeft(), data));
            if (2 == AvlUtil.height(root.getRight()) - AvlUtil.height(root.getLeft())) {
                if (root.getRight().getLeft() != null
                        && AvlUtil.height(root.getRight().getLeft()) > AvlUtil.height(root.getRight().getRight())) {
                    return doubleRotateRL(root);
                } else {
                    return singleRotateRight(root);
                }
            }
        } else if (root.getData() < data) {
            root.setRight(deleteNode(root.getRight(), data));
            if (2 == AvlUtil.height(root.getLeft()) - AvlUtil.height(root.getRight())) {
                if (root.getLeft().getRight() != null
                        && AvlUtil.height(root.getLeft().getRight()) > AvlUtil.height(root.getLeft().getLeft())) {
                    return doubleRotateLR(root);
                } else {
                    return singleRotateLeft(root);
                }
            }
        } else {
            if (root.getRight() != null && root.getLeft() != null) {
                // 找到右子数里最小的节点
                TreeNode right = root.getRight();
                while (right.getLeft() != null) {
                    right = right.getLeft();
                }

                root.setData(right.getData());
                root.setFreq(right.getFreq());

                root.setRight(deleteNode(root.getRight(), right.getData()));
                if (2 == AvlUtil.height(root.getLeft()) - AvlUtil.height(root.getRight())) {
                    if (root.getLeft().getRight() != null
                            && AvlUtil.height(root.getLeft().getRight()) > AvlUtil.height(root.getLeft().getLeft())) {
                        return doubleRotateLR(root);
                    } else {
                        return singleRotateLeft(root);
                    }
                }
            } else {
                if (root.getLeft() != null) {
                    root = root.getLeft();
                } else if (root.getRight() != null) {
                    root = root.getRight();
                } else {
                    root = null;
                }
            }
        }

        if (root != null) {
            root.reComputeHeight();
        }
        return root;
    }

    private void inOrderTraverse(TreeNode root) {
        if (root == null) {
            return;
        }

        inOrderTraverse(root.getLeft());
        System.out.println(root.getData());
        inOrderTraverse(root.getRight());
    }

    private TreeNode insertNode(TreeNode root, final int data) {
        if (root == null) {
            return new TreeNode(data);
        }

        if (root.getData() > data) {
            root.setLeft(insertNode(root.getLeft(), data));
            if (2 == AvlUtil.height(root.getLeft()) - AvlUtil.height(root.getRight())) {
                if (data < root.getLeft().getData()) {
                    return singleRotateLeft(root);
                } else {
                    return doubleRotateLR(root);
                }
            }
        } else if (root.getData() < data) {
            root.setRight(insertNode(root.getRight(), data));
            if (2 == AvlUtil.height(root.getRight()) - AvlUtil.height(root.getLeft())) {
                if (data > root.getRight().getData()) {
                    return singleRotateRight(root);
                } else {
                    return doubleRotateRL(root);
                }
            }
        } else {
            root.setFreq(root.getFreq() + 1);
        }

        root.reComputeHeight();
        return root;
    }

    // 右左情况下，先右旋，再左旋。这种旋转是，先让下面的节点旋转，再旋转上面的节点
    private TreeNode doubleRotateRL(TreeNode root) {
        root.setRight(singleRotateLeft(root.getRight()));
        return singleRotateRight(root);
    }

    // 右右情况下，向左旋转
    private TreeNode singleRotateRight(TreeNode root) {
        TreeNode newRoot = root.getRight();
        root.setRight(newRoot.getLeft());
        newRoot.setLeft(root);

        root.reComputeHeight();
        newRoot.reComputeHeight();
        return newRoot;
    }

    // 左右情况下，先左旋，后右旋。这种旋转是，先让下面的节点旋转，再旋转上面的节点
    private TreeNode doubleRotateLR(TreeNode root) {
        root.setLeft(singleRotateRight(root.getLeft()));
        return singleRotateLeft(root);
    }

    // 左左情况下，向右旋转
    private TreeNode singleRotateLeft(TreeNode root) {
        TreeNode newRoot = root.getLeft();
        root.setLeft(newRoot.getRight());
        newRoot.setRight(root);

        root.reComputeHeight();
        newRoot.reComputeHeight();
        return newRoot;
    }

    public static void main(String[] args) {
        AvlTree avlTree = new AvlTree();
        avlTree.insertData(5);
        avlTree.insertData(1);
        avlTree.insertData(3);
        avlTree.insertData(7);
        avlTree.insertData(9);

        avlTree.deleteData(5);

        avlTree.insertData(3);
        avlTree.insertData(11);
        avlTree.insertData(19);
        avlTree.insertData(12);
        avlTree.insertData(13);
        avlTree.insertData(14);
        avlTree.insertData(15);
        avlTree.insertData(15);
        avlTree.insertData(16);
        avlTree.insertData(17);
        avlTree.insertData(18);
        avlTree.insertData(19);
        avlTree.insertData(20);
        avlTree.insertData(21);
        avlTree.insertData(22);
        avlTree.insertData(23);
        avlTree.insertData(24);
        avlTree.insertData(25);

        avlTree.traverse();
    }
}
